tedge/src/system_services/systemd/mod.rs

```diff
@@ -0,0 +1,230 @@
+                service: service_name,
+                cmd: SystemCtlCmd::Restart.as_str(),
+                hint: "Lacking permissions or service's process exited with error code.",
+            }
+            .into()),
+            SYSTEMCTL_ERROR_SERVICE_NOT_FOUND => Err(SystemdError::ServiceNotFound {
+                service: service_name,
+            }
+            .into()),
+            code => Err(SystemdError::UnhandledReturnCode { code }.into()),
+        }
+    }
+
+    fn enable_service(&mut self, service: SystemService) -> Result<(), ServicesError> {
+        let service_name = service_name_for(service);
+        match self.call_systemd_subcmd_sudo(SystemCtlCmd::Enable, service_name)? {
+            SYSTEMCTL_OK => Ok(()),
+            SYSTEMCTL_ERROR_GENERIC => Err(SystemdError::UnspecificError {
+                service: service_name,
+                cmd: SystemCtlCmd::Enable.as_str(),
+                hint: "Lacking permissions.",
+            }
+            .into()),
+            code => Err(SystemdError::UnhandledReturnCode { code }.into()),
+        }
+    }
+
+    fn disable_service(&mut self, service: SystemService) -> Result<(), ServicesError> {
+        let service_name = service_name_for(service);
+        match self.call_systemd_subcmd_sudo(SystemCtlCmd::Disable, service_name)? {
+            SYSTEMCTL_OK => Ok(()),
+            SYSTEMCTL_ERROR_GENERIC => Err(SystemdError::UnspecificError {
+                service: service_name,
+                cmd: SystemCtlCmd::Disable.as_str(),
+                hint: "Lacking permissions.",
+            }
+            .into()),
+            code => Err(SystemdError::UnhandledReturnCode { code }.into()),
+        }
+    }
+
+    fn is_service_active(&mut self, service: SystemService) -> Result<bool, ServicesError> {
+        let service_name = service_name_for(service);
+        match self.call_systemd_subcmd(SystemCtlCmd::IsActive, service_name)? {
+            SYSTEMCTL_OK => Ok(true),
+            SYSTEMCTL_ERROR_UNIT_IS_NOT_ACTIVE => Ok(false),
+            code => Err(SystemdError::UnhandledReturnCode { code }.into()),
+        }
+    }
+}
+
+impl SystemdManager {
+    fn call_systemd_subcmd_sudo(
+        &self,
+        systemctl_subcmd: SystemCtlCmd,
+        arg: &str,
+    ) -> Result<i32, ServicesError> {
+        let _root_guard = self.user_manager.become_user(ROOT_USER);
+        self.call_systemd_subcmd(systemctl_subcmd, arg)
+    }
+
+    fn call_systemd_subcmd(
+        &self,
+        systemctl_subcmd: SystemCtlCmd,
+        arg: &str,
+    ) -> Result<i32, ServicesError> {
+        cmd_nullstdio_args_with_code(&self.systemctl_bin, &[systemctl_subcmd.as_str(), arg])?
+            .code()
+            .ok_or(ServicesError::UnexpectedExitStatus)
+    }
+}
+
+fn service_name_for(service: SystemService) -> &'static str {
+    match service {
+        SystemService::Mosquitto => "mosquitto",
+        SystemService::TEdgeMapper => "tedge-mapper",
+    }
+}
```

Comment:
This is independent of `systemd`. I would move it as a method of the `SystemServiceManager` trait.

Reply:
Good idea. Fixed

Reply:
I added it directly to the `SystemService` struct. Having this on the `SystemServiceManager` trait would also work pretty well and would allow for changing service name depending on the the manager.